typing-c/src/C/Typing/Clang.hs

```diff
@@ -0,0 +1,315 @@
+        Clang.CXType_Auto                -> do { canTy <- Clang.clang_getCanonicalType cxTy
+                                               ; parseClangType canTy }
+        Clang.CXType_Elaborated          -> do { namedTy <- Clang.clang_Type_getNamedType cxTy
+                                               ; parseClangType namedTy }
+
+-- | Query @clang@ for canonical names for types.
+getCanonicalTypeMapping :: [ Type ] -> IO ( Map Type Type )
+getCanonicalTypeMapping tys = do
+  cxTys <- Map.fromList <$> callClang ( getCanonicalType Nothing ) sourceProgram
+  traverse ( \ cxTy -> expectJust cxTy =<< parseClangType cxTy ) cxTys
+
+  where
+
+    getCanonicalType :: Maybe Int -> Clang.Fold IO ( Type, Clang.CXType )
+    getCanonicalType inTestFunDecl cursor = do
+      loc <- liftIO $ Clang.clang_getCursorLocation cursor
+      inMain <- liftIO $ Clang.clang_Location_isFromMainFile loc
+      if not inMain
+      then
+        return $ Clang.Continue Nothing
+      else do
+        cursorKind <- liftIO $ Clang.fromSimpleEnum <$> Clang.clang_getCursorKind cursor
+        case cursorKind of
+          Right kind
+            | Clang.CXCursor_FunctionDecl <- kind
+            -> do
+              mbFunNm <- liftIO $ Clang.getUserProvided <$> Clang.clang_getCursorSpelling cursor
+              return $
+                case mbFunNm of
+                  Just funNm
+                    | let ( nm, nb ) = Text.splitAt 6 funNm
+                    , nm == "testFn"
+                    , Just i <- readMaybe ( Text.unpack nb )
+                    -> Clang.Recurse ( getCanonicalType ( Just i ) ) listToMaybe
+                  _ ->
+                    Clang.Continue Nothing
+            | Just nb <- inTestFunDecl
+            , Clang.CXCursor_DeclRefExpr <- kind
+            -> do
+              cxTy   <- liftIO $ Clang.clang_getCursorType    cursor
+              return $ Clang.Continue ( Just ( tyPairs IntMap.! nb, cxTy ) )
+          _ -> return $ Clang.Recurse ( getCanonicalType inTestFunDecl ) listToMaybe
+
+    tyPairs :: IntMap Type
+    tyPairs = IntMap.fromList [ (i, ty) | i <- [ (1 :: Int) .. ] | ty <- tys ]
+
+    sourceProgram :: String
+    sourceProgram = unlines $ concat
+      [ [ "#include <stddef.h>" ]
+      , [ unlines
+            [ "static " ++ showTy "testFn" ++ show i ++ "(" ++ showTy "x" ++ ") {"
+            , "  return x;"
+            , "}"
+            ]
+        | ( i, ty ) <- IntMap.assocs tyPairs
+        , let showTy = showTypeAsCType ty
+        ]
+      ]
+
+    expectJust :: Clang.CXType -> Maybe a -> IO a
+    expectJust cxTy =
+      \case
+        Nothing -> do
+          tyNm <- Clang.clang_getTypeSpelling cxTy
+          error $ unlines
+            [ "getCanonicalTypeMapping: could not parse CXType " ++ show cxTy
+            , Text.unpack tyNm ]
+        Just ty -> return ty
+
+queryClangForResultType :: forall n. Vec n Type -> ( Vec n String -> String ) -> IO ( Maybe Type )
```

Comment:
Can we use `Clang.CXType` instead of `String` for the `Vec n String` argument please?